enterprise/internal/codeintel/shared/resolvers/dataloader.go

```diff
@@ -0,0 +1,117 @@
+package sharedresolvers
+
+import (
+	"context"
+	"sync"
+)
+
+type Identifiable[T any] interface {
+	GetID() T
+}
+
+type MultiFactory[K, V any] interface {
+	Create(ctx context.Context, id K) ([]V, error)
+}
+
+type MultiFactoryFunc[K, V any] func(ctx context.Context, id K) ([]V, error)
+
+func (f MultiFactoryFunc[K, V]) Create(ctx context.Context, id K) ([]V, error) {
+	return f(ctx, id)
+}
+
+type DoubleLockedCache[K comparable, V Identifiable[K]] struct {
+	sync.RWMutex
+	factory MultiFactory[K, V]
+	cache   map[K]V
+}
+
+func NewDoubleLockedCache[K comparable, V Identifiable[K]](factory MultiFactory[K, V]) *DoubleLockedCache[K, V] {
+	return &DoubleLockedCache[K, V]{
+		factory: factory,
+		cache:   map[K]V{},
+	}
+}
+
+func (c *DoubleLockedCache[K, V]) GetOrCreate(ctx context.Context, id K) (obj V, ok bool, _ error) {
+	c.RLock()
+	obj, ok = c.cache[id]
+	c.RUnlock()
+	if ok {
+		return obj, true, nil
+	}
+
+	c.Lock()
+	defer c.Unlock()
+
+	if obj, ok := c.cache[id]; ok {
+		return obj, true, nil
+	}
+
+	objs, err := c.factory.Create(ctx, id)
+	if err != nil {
+		return obj, false, err
+	}
+	for _, obj := range objs {
+		c.cache[obj.GetID()] = obj
+	}
+
+	obj, ok = c.cache[id]
+	return obj, ok, nil
+}
+
+type DataLoader[K comparable, V Identifiable[K]] struct {
+	svc   DataLoaderBackingService[K, V]
+	ids   map[K]struct{}
+	cache *DoubleLockedCache[K, V]
+}
+
+type DataLoaderBackingService[K comparable, V Identifiable[K]] interface {
+	GetByIDs(ctx context.Context, ids ...K) ([]V, error)
+}
+
+type DataLoaderBackingServiceFunc[K, V any] func(ctx context.Context, ids ...K) ([]V, error)
+
+func (f DataLoaderBackingServiceFunc[K, V]) GetByIDs(ctx context.Context, ids ...K) ([]V, error) {
+	return f(ctx, ids...)
+}
+
+func NewDataLoader[K comparable, V Identifiable[K]](svc DataLoaderBackingService[K, V]) *DataLoader[K, V] {
+	dl := &DataLoader[K, V]{
+		svc: svc,
+		ids: map[K]struct{}{},
+	}
+
+	dl.cache = NewDoubleLockedCache[K, V](MultiFactoryFunc[K, V](dl.load))
+	return dl
+}
+
+func (l *DataLoader[K, V]) Mark(ids ...K) {
```

Comment:
Mark who?

Find the name a bit non-self-explanatory, so maybe try something like or whatever

```suggestion
func (l *DataLoader[K, V]) Presubmit(ids ...K) {
```